File: src/configs.ts
import { Properties } from "./properties";
import { Display, DISPLAYS, OperatingSystem, OPERATING_SYSTEMS } from "./state";

type ConfigProviderOperatingSystem = 'Linux' | 'Windows';

export class ConfigurationError extends Error {
  constructor(message: string, configProviderOS: ConfigProviderOperatingSystem) {
    super(`${message}. Execute 'my-reboot configure' no ${configProviderOS}`);
    Object.setPrototypeOf(this, ConfigurationError.prototype);
  }
}

class ConfigHandler<T> {
  // object.attribute = value
  private readonly props: Properties;
  private readonly attribute: string;
  private readonly configProviderOS: ConfigProviderOperatingSystem;
  private readonly objects: ReadonlyArray<T>;

  constructor(props: Properties, attribute: string, objects: ReadonlyArray<T>, configProviderOS: ConfigProviderOperatingSystem) {
    this.props = props;
    this.attribute = attribute;
    this.objects = objects;
    this.configProviderOS = configProviderOS;
  }

  getValue(object: T): string {
    const key = this.keyFor(object);
    const value = this.props.get(key);
    if (value !== undefined) {
      return value;
    } else {
      throw new ConfigurationError(`Configuração '${key}' não encontrada`, this.configProviderOS);
    }
  }

  getObjectByValue(value: string): T {
    const obj = this.objects.find(obj => this.getValue(obj) === value);
    if (obj !== undefined) {
      return obj;
    } else {
      throw new ConfigurationError(`Configuração com valor ${value} não encontrada`, this.configProviderOS);
    }
  }

  setValue(object: T, value: string) {
    const key = this.keyFor(object);
    this.props.set(key, value);
  }

  private keyFor(object: T) {
    return `${object}.${this.attribute}`;
  }
}

export class Configs {
  protected readonly grubEntryHandler: ConfigHandler<OperatingSystem>;
  protected readonly deviceIdHandler: ConfigHandler<Display>;
  protected readonly displaySwitchArgHandler: ConfigHandler<Display>;

  protected static readonly FILE = 'my-reboot-configs.properties';

  static async load(stateDir: string): Promise<Configs> {
    return await ConfigsWriter.load(stateDir);
  }

  protected constructor(props: Properties) {
    this.grubEntryHandler = new ConfigHandler<OperatingSystem>(props, 'grubEntry', OPERATING_SYSTEMS, 'Linux');
    this.deviceIdHandler = new ConfigHandler<Display>(props, 'deviceId', DISPLAYS, 'Windows');
    this.displaySwitchArgHandler = new ConfigHandler<Display>(props, 'displaySwitchArg', DISPLAYS, 'Windows');
  }

  getOperatingSystemByGrubEntry(grubEntry: string): OperatingSystem {
    return this.grubEntryHandler.getObjectByValue(grubEntry);
  }

  getGrubEntry(operatingSystem: OperatingSystem): string {
    return this.grubEntryHandler.getValue(operatingSystem);
  }

  getDisplayByDeviceId(deviceId: string): Display {
    return this.deviceIdHandler.getObjectByValue(deviceId);
  }

  getDeviceId(display: Display): string {
    return this.deviceIdHandler.getValue(display);
  }

  getDisplaySwitchArg(display: Display): string {
    return this.displaySwitchArgHandler.getValue(display);
  }
}

export class ConfigsWriter extends Configs {
  static override async load(stateDir: string): Promise<ConfigsWriter> {
    const props = await Properties.load(`${stateDir}/${Configs.FILE}`);
    return new ConfigsWriter(props);
  }

  private readonly props: Properties;

  private constructor(props: Properties) {
    super(props);
    this.props = props;
  }

  setGrubEntry(operatingSystem: OperatingSystem, value: string) {
    this.grubEntryHandler.setValue(operatingSystem, value);
  }

  setDeviceId(display: Display, value: string) {
    this.deviceIdHandler.setValue(display, value);
  }

  setDisplaySwitchArg(display: Display, value: string) {
    this.displaySwitchArgHandler.setValue(display, value);
  }

  async save() {
    await this.props.save();
  }
}

File: src-tauri/src/configs.rs
use std::io;
use std::rc::Rc;

#[cfg(windows)]
use crate::options_types::Display;
use crate::options_types::{OperatingSystem, OptionType};
use crate::properties::Properties;

const CONFIGS_FILENAME: &str = "my-reboot-configs.properties";

pub struct Configs {
    grub_entry_handler: ConfigHandler<OperatingSystem>,
    #[cfg(windows)]
    device_id_handler: ConfigHandler<Display>,
    #[cfg(windows)]
    display_switch_arg_handler: ConfigHandler<Display>,
}

impl Configs {
    pub fn load(must_exist: bool) -> io::Result<Configs> {
        let props = Properties::load(CONFIGS_FILENAME, must_exist)?;
        let props = Rc::new(props);

        Ok(Configs {
            grub_entry_handler: ConfigHandler::new(
                Rc::clone(&props),
                "grubEntry",
                OperatingSystem::values(),
                OperatingSystem::Linux,
            ),
            #[cfg(windows)]
            device_id_handler: ConfigHandler::new(
                Rc::clone(&props),
                "deviceId",
                Display::values(),
                OperatingSystem::Windows,
            ),
            #[cfg(windows)]
            display_switch_arg_handler: ConfigHandler::new(
                Rc::clone(&props),
                "displaySwitchArg",
                Display::values(),
                OperatingSystem::Windows,
            ),
        })
    }

    pub fn get_operating_system_by_grub_entry(&self, grub_entry: &str) -> OperatingSystem {
        self.grub_entry_handler.get_object_by_value(grub_entry)
    }

    pub fn get_grub_entry(&self, os: OperatingSystem) -> &str {
        self.grub_entry_handler.get_value(os)
    }

    #[cfg(windows)]
    pub fn get_display_by_device_id(&self, device_id: &str) -> Display {
        self.device_id_handler.get_object_by_value(device_id)
    }

    #[cfg(windows)]
    pub fn get_display_switch_arg(&self, display: Display) -> &str {
        self.display_switch_arg_handler.get_value(display)
    }
}

struct ConfigHandler<O: OptionType> {
    props: Rc<Properties>,
    attribute: &'static str,
    objects: Vec<O>,
    config_provider_os: OperatingSystem,
}
impl<O: OptionType + Copy> ConfigHandler<O> {
    fn new(
        props: Rc<Properties>,
        attribute: &'static str,
        objects: Vec<O>,
        config_provider_os: OperatingSystem,
    ) -> ConfigHandler<O> {
        ConfigHandler {
            props,
            attribute,
            objects,
            config_provider_os,
        }
    }

    fn get_object_by_value(&self, value: &str) -> O {
        let object = self
            .objects
            .iter()
            .find(|o| self.get_value(**o) == value)
            .unwrap_or_else(|| {
                panic!(
                    "{}",
                    configuration_error(
                        &format!("Configuração com valor {value} não encontrada"),
                        self.config_provider_os,
                    )
                )
            });
        *object
    }

    fn get_value(&self, object: O) -> &str {
        let key = self.key_for(object);
        self.props.get(&key).unwrap_or_else(|| {
            panic!(
                "{}",
                configuration_error(
                    &format!("Configuração '{key}' não encontrada"),
                    self.config_provider_os,
                )
            )
        })
    }

    fn key_for(&self, object: O) -> String {
        format!("{}.{}", object.to_option_string(), self.attribute)
    }
}

fn configuration_error(message: &str, config_provider_os: OperatingSystem) -> String {
    format!(
        "{message}. Execute 'my-reboot configure' no {}",
        config_provider_os
    )
}
